### PR TITLE
Filter instances on backend side

### PR DESCRIPTION
**Pull request by krhubert**
clsoes https://github.com/mesg-foundation/engine/issues/1467

### CODE REVIEW

File: sdk/instance/backend.go [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package instancesdk
 
 import (
+	"encoding/json"
```
Comment (NicolasMahe) [inline sdk/instance/backend.go]: use codec instead of json

File: sdk/instance/backend.go [hunk 2]
```diff
@@ -46,7 +48,12 @@ func (s *Backend) querier(request cosmostypes.Request, path []string, req abci.R
 		}
 		return s.Get(request, hash)
 	case "list":
-		return s.List(request)
+		var f api.ListInstanceRequest_Filter
+		if err := json.Unmarshal(req.Data, &f); err != nil {
```
Comment (NicolasMahe) [inline sdk/instance/backend.go]: could you move the unmarshal in:
https://github.com/mesg-foundation/engine/blob/940abdaa67779f39482e03b0bb6afbac0a05b745/cosmos/module.go#L112-L127

this function already marshal the data before sending it back

File: sdk/instance/sdk.go [hunk 3]
```diff
@@ -35,26 +38,17 @@ func (s *SDK) Get(hash hash.Hash) (*instance.Instance, error) {
 }
 
 // List returns all instances.
-func (s *SDK) List(f *Filter) ([]*instance.Instance, error) {
-	var instances []*instance.Instance
-	if err := s.client.Query("custom/"+backendName+"/list", nil, &instances); err != nil {
+func (s *SDK) List(f *api.ListInstanceRequest_Filter) ([]*instance.Instance, error) {
```
Comment (NicolasMahe) [inline sdk/instance/sdk.go]: don't forget to delete sdk.Filter line 15

File: sdk/instance/sdk.go [hunk 4]
```diff
@@ -35,26 +38,17 @@ func (s *SDK) Get(hash hash.Hash) (*instance.Instance, error) {
...
+	fb, err := json.Marshal(f)
```
Comment (NicolasMahe) [inline sdk/instance/sdk.go]: could you move the marshal inside the client.Query. it already unmarshal the received data.

File: sdk/instance/backend.go (hunk 2 above)
Comment (NicolasMahe) [inline sdk/instance/backend.go]: should use codec

**Comment by krhubert** [comment]
I don't know why e2e fails, can you help me fix that?

File: cosmos/client.go [hunk 5]
```diff
@@ -34,8 +34,8 @@ func NewClient(node *node.Node, kb keys.Keybase, chainID string) *Client {
 }
 
 // Query is abci.query wrapper with errors check and decode data.
-func (c *Client) Query(path string, data cmn.HexBytes, ptr interface{}) error {
-	result, err := c.ABCIQuery(path, data)
+func (c *Client) Query(path string, data, ptr interface{}) error {
+	result, err := c.ABCIQuery(path, codec.MustMarshalJSON(data))
```
Comment (NicolasMahe) [inline cosmos/client.go]:
```suggestion
	dataB, err := codec.MarshalJSON(data)
	if err != nil {
		return err
	}
	result, err := c.ABCIQuery(path, dataB)
```

**Comment by NicolasMahe** [comment]
> I don't know why e2e fails, can you help me fix that?

`servicesdk.hash` is using `proto.Marshal`.
```
// Hash returns the calculate hash of a service.
func (s *SDK) Hash(req *api.CreateServiceRequest) (hash.Hash, error) {
	var h hash.Hash
	b, err := proto.Marshal(req)
	if err != nil {
		return nil, err
	}
	if err := s.client.Query("custom/"+backendName+"/hash", cmn.HexBytes(b), &h); err != nil {
		return nil, err
	}
	return h, nil
}
```

**Comment by NicolasMahe** [comment]
> > I don't know why e2e fails, can you help me fix that?
> 
> `servicesdk.hash` is using `proto.Marshal`.
> 
> ```
> // Hash returns the calculate hash of a service.
> func (s *SDK) Hash(req *api.CreateServiceRequest) (hash.Hash, error) {
> 	var h hash.Hash
> 	b, err := proto.Marshal(req)
> 	if err != nil {
> 		return nil, err
> 	}
> 	if err := s.client.Query("custom/"+backendName+"/hash", cmn.HexBytes(b), &h); err != nil {
> 		return nil, err
> 	}
> 	return h, nil
> }
> ```

I fixed it ;)

File: sdk/service/backend.go [hunk 6]
```diff
@@ -66,7 +65,7 @@ func (s *Backend) querier(request cosmostypes.Request, path []string, req abci.R
 		return s.List(request)
 	case "hash":
 		var createServiceRequest api.CreateServiceRequest
-		if err := proto.Unmarshal(req.Data, &createServiceRequest); err != nil {
+		if err := codec.UnmarshalJSON(req.Data, &createServiceRequest); err != nil {
```
Comment (antho1404) [inline sdk/service/backend.go]: I would prefer to use proto for the encoding here as we had issues with the json encoding when using `oneof`. For the service it is fine but if we do that for the process we will have an issue. That was the reason why we used proto to marshal here and at the same time it's more efficient

File: sdk/instance/backend.go [hunk 7]
```diff
@@ -91,18 +97,21 @@ func (s *Backend) Exists(request cosmostypes.Request, hash hash.Hash) (bool, err
...
 	)
 
+	// filter results
 	for iter.Valid() {
 		var i *instance.Instance
 		if err := codec.UnmarshalBinaryBare(iter.Value(), &i); err != nil {
 			return nil, err
 		}
-		instances = append(instances, i)
+		if f == nil || (f != nil && (f.ServiceHash.IsZero() || i.ServiceHash.Equal(f.ServiceHash))) {
```
Comment (antho1404) [inline sdk/instance/backend.go]: `f` will not be nil if we pass the first 'or', also inverted the last condition to be more consistent with the conditions

```suggestion
		if f == nil || f.ServiceHash.IsZero() || f.ServiceHash.Equal(i.ServiceHash) {
```

File: sdk/service/backend.go (hunk 6 above)
Comment (NicolasMahe) [reply sdk/service/backend.go]: we have to choose one encoding for the data. we cannot use proto and amino and json.
Service is not using any oneof, so what is the issue here?

Comment (antho1404) [reply sdk/service/backend.go]: I'm fine to use binary encode from amino, let's just try to avoid json

Comment (NicolasMahe) [reply sdk/service/backend.go]: @krhubert can you replace by binary unmarshal? don't forget to replace also the marshal ;)

File: sdk/instance/codec.go [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package instancesdk
+
+import (
+	"github.com/mesg-foundation/engine/codec"
+	"github.com/mesg-foundation/engine/protobuf/api"
+)
+
+func init() {
+	codec.Codec.RegisterConcrete(&api.ListInstanceRequest_Filter{}, "mesg.types.ListInstanceRequest_Filter ", nil)
```
Comment (NicolasMahe) [inline sdk/instance/codec.go]: why do you need to register ListInstanceRequest_Filter to the codec? I tried without and its working well.

File: sdk/service/codec.go [hunk 9]
```diff
@@ -2,8 +2,10 @@ package servicesdk
 
 import (
 	"github.com/mesg-foundation/engine/codec"
+	"github.com/mesg-foundation/engine/service"
 )
 
 func init() {
 	codec.RegisterConcrete(msgCreateService{}, "service/create", nil)
+	codec.RegisterConcrete(&service.Service{}, "mesg.types.Service", nil)
```
Comment (NicolasMahe) [inline sdk/service/codec.go]: why do you need to register service to the codec? I tried without and its working well.
Only struct used inside an interface should be registered. The msgs have to be register as they are used in Tx as interface

**Comment by krhubert** [comment]
@antho1404 

Fixed issues with panic. Also add e2e tests for different instance hash and it works as expected. I'm not going to debug grpcurl as it is just a helper tool. From e2e tests and code perespective everything work fine.

**Pull request merged**